OutOfSchool/OutOfSchool.Redis/MultiLayerCache.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OutOfSchool.Redis;
+
+public sealed class MultiLayerCache : IMultiLayerCacheService, IDisposable
+{
+    private readonly ReaderWriterLockSlim cacheLock = new ReaderWriterLockSlim();
+    private readonly IMemoryCache _memoryCache;
+    private readonly ICacheService _cacheService;
+    private readonly MemoryCacheConfig memoryCacheConfig;
+
+    public MultiLayerCache(
+        ICacheService externalCache,
+        IMemoryCache inMemoryCache,
+        IOptions<MemoryCacheConfig> memoryCacheConfig
+    )
+    {
+        this._memoryCache = inMemoryCache;
+        this._cacheService = externalCache;
+        this.memoryCacheConfig = memoryCacheConfig.Value;
+    }
+
+    public Task ClearCacheAsync(string key)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+            try
+            {
+                _memoryCache.Remove(key);
+                _cacheService.ClearCacheAsync(key);
```

Comment:
`_cacheService.ClearCacheAsync(key);` - is never awaited. 
So your `ClearCacheAsync()` implementation could return "completed" task earlier than it is really completed.

Reply:
Same for `Refresh()`

Reply:
Added awaiters.